File: src/654_Maximum_Binary_Tree/main.cpp

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

/**
 Definition for a binary tree node.*/
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    //Runtime: 112 ms, faster than 44.93%
    //Memory Usage: 42.8 MB, less than 32.88%
    TreeNode* constructMaximumBinaryTree1(vector<int>& nums) {
        // 思路是转化为子问题，然后拆分为两个子树分别处理
        // 在考量能否改进一点的地方是增加一个stack，用于来存储已经找到的最大值，这样避免重复搜索
        
        return helper(nums,0,nums.size());// 复杂度对于[1,1000]应该搞得定，大不了是O(n^2)
    }
    
    TreeNode* helper(vector<int>& nums, int bgn, int end) {
        int maxPos = bgn-1, maxVal = INT_MIN;
        // if (bgn < 0 || bgn >= nums.size()) return NULL; 大意了大意了啊！！！
        if (bgn < 0 || bgn >= end) return NULL;
        
        for (int i = bgn; i < end; ++i) {
            // 因为全部是unique值，所以没有等于的情况
            if (maxVal < nums[i]) {
                maxVal = nums[i];
                maxPos = i;
            }
        }
        
        TreeNode* root = new TreeNode(nums[maxPos]);
        TreeNode* l_node = helper(nums, bgn, maxPos);
        TreeNode* r_node = helper(nums, maxPos+1, end);
        
        root->left = l_node; root->right = r_node;
        return root;
    }
    
    // 另一种做法，非常巧妙：就是使用了stack来简化for遍历的操作，有点像中序遍历的迭代方法
    //Runtime: 108 ms, faster than 49.78% 
    //Memory Usage: 39.6 MB, less than 43.80%
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        vector<TreeNode*> stk;
        for (int i = 0; i < nums.size(); ++i) {
            TreeNode* cur = new TreeNode(nums[i]);
            // 下面这个while循环是为了找到第一个比当前cur大的node，
            // 如果找到；那么cur必定属于这个node的右节点，因为node会分割左右；
            // 同时这个cur又会是那些比cur还小的node的父节点
            while (!stk.empty() && stk.back()->val < nums[i]) {
                cur->left = stk.back(); // 因为下面会被pop，所以先存下来；会一直更新到比cur小的那个最大的node
                stk.pop_back();
            }
            // 跳出while循环说明stack里面存的node都比当前node大（或者stack为空）
            // 对于栈不为空的情况，按照题意最大值的node为分割，那么当前node必定是之前那个node的右节点
            
            if (!stk.empty())
                stk.back()->right = cur;
            
            stk.push_back(cur);//为了防止下一个node比cur小，需要将cur也存入stack
        }
        return stk.front();
    }
};

ostream & operator<<(ostream &os, TreeNode* root) {
    os << "[";
    deque<TreeNode*> helper{root};
    while (helper.size()) {
        int size = helper.size();
        
        for (int i = 0; i < size; ++i) {
            TreeNode* tmp = helper.front();
            helper.pop_front();
            
            os << tmp->val << ",";
            if (tmp->left) helper.push_back(tmp->left);
            if (tmp->right) helper.push_back(tmp->right);
        }
    }
    
    os << "]";
    return os;
}

int main(int argc, const char * argv[]) {
    Solution gua;
    TreeNode* rtn;
    
    vector<int> nums{3,2,1,6,0,5};
    rtn = gua.constructMaximumBinaryTree(nums);
    cout << rtn << endl;

    return 0;
}

File: src/654_Maximum_Binary_Tree/main.py
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    pos = 0
    # 原本打算硬算的，就是遍历到最大值，然后分两段，再分别对两段进行递归，但是复杂度是O(n*logn)
    # 不过观察到两个关键点，1，root是最大值，2，新加入的点总是再右手边，于是感觉可以O(n)解决
    # Runtime: 204 ms, faster than 66.45%
    # Memory Usage: 15 MB, less than 27.47%
    def appendToTree(self, root: TreeNode, nums: List[int]) -> TreeNode:
        # 返回条件
        if self.pos >= len(nums):
            return root
        elif not root:
            root = TreeNode(nums[self.pos])
            self.pos = self.pos+1
            #return appendToTree(root, nums) 不能就这么返回，继续做操作
            return root
            # if self.pos >= len(nums)-1:
            #     return root
            
        # 现在root和pos都有效，与root进行判断，因为题目说val是unique的，所以很开心
        #root.right = appendToTree(root.right, pos, nums)
        if root.val < nums[self.pos]:
            # 先处理简单情况
            newRoot = TreeNode(nums[self.pos])
            self.pos = self.pos+1
            #newRoot.left = root 先不要武断的添加，交由上层appendToTree进行处理
            return newRoot
            
        # 由于可能出现后加的val大的情况，所以还不能武断的加入root.right
        tmp = self.appendToTree(root.right, nums)
        if root.val > tmp.val:
            if not root.right:
                root.right = tmp
            elif root.right != tmp:
                # 说明tmp比root的右分支大
                # 那么tmp的左分支继承原本的右分支
                tmp.left = root.right #tmp.right = root.right
                # 再把原本的右分支替换为tmp
                root.right = tmp

            # root.right = tmp if not root.right else root.right
            # 仅仅在root的右分支与新节点不同时，更新root,小于则不做处理
            # if root.right != tmp:
            #     tmp.right = root.right
            #     root.right = tmp
            # return root # 直接返回root可能还没有算完
            return self.appendToTree(root, nums)
        else:
            return tmp
        # if root.val > nums[pos]:
        #     # 比root小，并且比root的顺序靠后
        # else:
        #     # 比root大，那么需要进行节点交换
        #     tmp = TreeNode(nums[pos])
        #     tmp.left = root
        #     return tmp
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        self.pos = 0 # python貌似没有reference，那么就用全局变量吧
        dummy = TreeNode(1001) #比所有节点都大；如果不告知这个nums的范围的话，那么就只能传入None，由函数内部自己判断
        root = self.appendToTree(dummy, nums)
        return dummy.right

def printNode(root: TreeNode):
  print("[",end="")

  st = [root]
  while len(st):
    size = len(st)
    level = ""
    isValid = False
    for i in range(size):
      cur = st[0]
      st.pop(0) # 总是从前面取; 默认不带参数的话等价于-1，从尾部取

      if cur:
        isValid = True
        level+=(str(cur.val)+",")
        if cur.left or cur.right:
          st.append(cur.left)
          st.append(cur.right)
      else:
        level+="null,"

    if isValid:
      print(level, end="")

  print("]")

if __name__ == "__main__":
  gua = Solution()
  nums = [3,2,1,6,0,5]
  rtn = gua.constructMaximumBinaryTree(nums)
  printNode(rtn)

  nums = [3,1,2]
  rtn = gua.constructMaximumBinaryTree(nums)
  printNode(rtn)

  nums = [3,2,1]
  rtn = gua.constructMaximumBinaryTree(nums)
  printNode(rtn)


